lib/kube/kubeconfig/kubeconfig.go

```diff
@@ -359,9 +367,46 @@ func SelectContext(teleportCluster, kubeCluster string) error {
 	if _, ok := kc.Contexts[kubeContext]; !ok {
 		return trace.NotFound("kubeconfig context %q not found", kubeContext)
 	}
+	setSelectedExtension(kc.Contexts, kc.CurrentContext, teleportCluster)
 	kc.CurrentContext = kubeContext
 	if err := Save("", *kc); err != nil {
 		return trace.Wrap(err)
 	}
 	return nil
 }
+
+const selectedExtension = "selected_context"
+
+// setSelectedExtension sets an extension to indentify that the current non-teleport
+// context was selected before introducing Teleport contexts in kubeconfig.
+// If the currentContext is not from Teleport, this function adds the following
+// extensions:
+//   - extension: null
+//     name: selected_context
+//
+// Only one context is allowed to have the selected extension. If other context has it,
+// this function deletes it and introduces it in the desired context.
+func setSelectedExtension(contexts map[string]*clientcmdapi.Context, prevCluster string, teleportCluster string) {
+	selected, ok := contexts[prevCluster]
+	if !ok || strings.HasPrefix(prevCluster, teleportCluster) || len(prevCluster) == 0 {
+		return
+	}
+	for _, v := range contexts {
+		delete(v.Extensions, selectedExtension)
+	}
+
+	selected.Extensions[selectedExtension] = nil
+}
+
+// searchForSelectedCluster looks for contexts that were previously selected
+// in order to restore the the CurrentContext value.
+// If no such key is found, it returns an empty selected cluster.
+func searchForSelectedCluster(contexts map[string]*clientcmdapi.Context) string {
+	for k, v := range contexts {
+		if _, ok := v.Extensions[selectedExtension]; ok {
+			delete(v.Extensions, selectedExtension)
+			return k
+		}
+	}
+	return ""
```

Comment:
I understand that `setSelectedExtension` flushes all other `selected_context` extensions, but users are sometimes editing/merging kubeconfigs and might end up with the extension set twice. Should this case be handled?

Reply:
Yeah, that's a good case. I will change it to return a non-empty current context only if a single cluster has the selected extension.